data_utils.py

```diff
@@ -0,0 +1,208 @@
+from torch.nn.utils.rnn import pad_sequence
+from torchtext.vocab import Vocab
+from torchtext.data.utils import get_tokenizer
+
+from utils import log
+
+
+class TextDataset(Dataset):
+    """Class for text dataset"""
+
+    def __init__(self, data, word_dict, classes, max_seq_length):
+        self.data = data
+        self.word_dict = word_dict
+        self.classes = classes
+        self.max_seq_length = max_seq_length
+        self.num_class = len(self.classes)
+        self.label_binarizer = MultiLabelBinarizer().fit([classes])
+
+    def __len__(self):
+        return len(self.data)
+
+    def __getitem__(self, index):
+        data = self.data[index]
+        return {
+            'text': torch.LongTensor([self.word_dict[word] for word in data['text']][:self.max_seq_length]),
+            'label': torch.FloatTensor(self.label_binarizer.transform([data['label']])[0]),
+            'index': data.get('index', 0)
+        }
+
+
+def generate_batch(data_batch):
+    text_list = [data['text'] for data in data_batch]
+    label_list = [data['label'] for data in data_batch]
+    return {
+        'index': [data['index'] for data in data_batch],
+        'text': pad_sequence(text_list, batch_first=True),
+        'label': torch.stack(label_list)
+    }
+
+
+def get_dataset_loader(config, dataset, shuffle=False, train=True):
+    dataset_loader = torch.utils.data.DataLoader(
+        dataset,
+        batch_size=config.batch_size if train else config.eval_batch_size,
+        shuffle=shuffle,
+        num_workers=config.data_workers,
+        collate_fn=generate_batch,
+        pin_memory='cuda' in config.device.type,
+    )
+    return dataset_loader
+
+
+def _load_raw_data(config, data_dir):
+    raw_data_cache = config.get('raw_data_cache')
+    if raw_data_cache and os.path.exists(raw_data_cache):
+        log.info(f'Load existing raw data cache from {raw_data_cache}.')
+        with open(raw_data_cache, 'rb') as fp:
+            dataset = pickle.load(fp)
+        if 'labels' in next(iter(dataset.values())):
+            for split in dataset:
+                dataset[split]['label'] = dataset[split].pop('labels')
+        return dataset
+
+    log.info(f'Load data from train_texts.txt, train_labels.txt, test_texts.txt, and text_labels.txt.')
+    dataset = {k: {} for k in ['train', 'test']}
+    for split in ['train', 'test']:
+        with open(os.path.join(data_dir, f'{split}_texts.txt')) as f:
+            texts = f.readlines()
+            dataset[split]['text'] = [text for text in texts]
+        with open(os.path.join(data_dir, f'{split}_labels.txt')) as f:
+            labels = f.readlines()
+            dataset[split]['label'] = [label.split() for label in labels]
+        dataset[split]['index'] = list(range(len(dataset[split]['text'])))
+    return dataset
+
+
+@log.enter('load_dataset')
+def load_dataset(config):
+    """Preparing TextDataset from raw data."""
+    data_dir = os.path.join(config['data_dir'], config['data_name'])
```

Comment:
Unify all `config['xxx'] syntax to `config.xxx`?

Reply:
Some hyperparameter search tools may require configs as a dictionary.
Correct me if I'm wrong.